[99][sig-policy] Policy placement strategy (#99)
open-cluster-management-io/enhancements: enhancements/sig-policy/99-policy-placement-strategy/README.md

### Goals

- Make `Placement` the primary API for placement (currently the governance propagator is somewhat
  `PlacementRule`-centric)
- Leverage the `Placement` helper library to retrieve placement decisions
- Reflect rollout status per cluster in the root policy for discoverability (whether a cluster is up-to-date or not)
- (stretch) Implement the `RolloutStrategy` struct for policies, including:
  - `RolloutStrategy`
    - "All": all clusters at once
    - "Progressive": one cluster at a time
    - "ProgressivePerGroup": one group of clusters at a time
  - `Timeout`: Maximum amount of time to wait for success before continuing the rollout
  - `MandatoryDecisionGroups`: groups that should be handled before other groups
  - `MaxConcurrency`: Concurrency during a progressive rollout
- (stretch) Add an aggregated rollout status for the root policy status.

### Non-Goals

Any specialized behaviors outside of those provided by the `Placement` library and, by extension, the `DecisionStrategy`
enhancement, that might require additional code other than that already provided/required by the `Placement` library.

> 

Reviewer: Maybe list Rollback here specifically.  Rollback would require the user to re-apply the old policy after seeing the canary failure.

Author: Added:
> This includes the ability to roll back a rollout. A rollback should be implemented by the user by applying the previous version of the policy, and GitOps is the recommended method for this.

## Summary

Following from the [`DecisionStrategy`](../../sig-architecture/64-placementStrategy/README.md) field in the `Placement`
API, policies can leverage this new logic to have a configurable and systematic way to roll out policy updates to
clusters.

## Motivation

Currently policies and subsequent updates are pushed out to clusters en masse based on the placement to which it has
been bound. The new `DecisionStrategy` field in the `Placement` API creates cluster groupings that controllers can
leverage to have segmented, configurable rollouts of resources. This will aid use cases where high availability may be a
priority or a set of test clusters should receive and verify the updates before the remaining clusters get updated.

### Goals

- Make `Placement` the primary API for placement (currently the governance propagator is somewhat
  `PlacementRule`-centric)
- Leverage the `Placement` helper library to retrieve placement decisions
- Reflect rollout status per cluster in the root policy for discoverability (whether a cluster is up-to-date or not)
- (stretch) Implement the `RolloutStrategy` struct for policies, including:
  - `RolloutStrategy`

>     - "All": all clusters at once

Reviewer 2: Would this be considered stretch because it's the current behavior? So at a minimum, "All" would be supported.

Author: Correct. At a minimum, we'd implement the `Placement` library using `All` without making it configurable by the user.

    - "Progressive": one cluster at a time
    - "ProgressivePerGroup": one group of clusters at a time
  - `Timeout`: Maximum amount of time to wait for success before continuing the rollout
  - `MandatoryDecisionGroups`: groups that should be handled before other groups

>   - `MaxConcurrency`: Concurrency during a progressive rollout

Reviewer 3: When `RolloutStrategy: Progressive`, MaxConcurrency isn't needed?

Author: Per the decision strategy struct, `MaxConcurrency` is only defined for `Progressive` rollouts: https://github.com/open-cluster-management-io/api/blob/2b2399b5f6e88e28a962aa82307ccc00bc492009/cluster/v1alpha1/types_rolloutstrategy.go#L95-L99

- (stretch) Add an aggregated rollout status for the root policy status.

### Non-Goals

Any specialized behaviors outside of those provided by the `Placement` library and, by extension, the `DecisionStrategy`
enhancement, that might require additional code other than that already provided/required by the `Placement` library.
This includes the ability to roll back a rollout. A rollback should be implemented by the user by applying the previous
version of the policy, and GitOps is the recommended method for this.

## Proposal

### User Stories

#### Story 1 / Phase 1

As a system administrator, I want to know the status of a rollout for a policy and what clusters have been updated.

- **Summary**

>   - Add an annotation to the replicated Policy indicating whether a policy is up-to-date.

Author: No--it's mainly implementing the metadata. I meant to put Story 1 and Story 2 both in Phase 1 since they could be completed concurrently and apparently neglected to do so. 🙂

- **Details**

>   - Update the `governance-policy-propagator` controller to add a `open-cluster-management.io/root-policy-generation`

Reviewer 2: This alone isn't sufficient because it doesn't take into account if the managed cluster acted upon it.

Author: Right--that's the purpose of the `lastEvaluated` status update in Phase 3. I was thinking for Phase 1 and 2, match current behaviors without figuring in other rollout strategies other than `All` so that the annotations could be implemented separately from the `Placement` implementation.

Reviewer 2: @dhaiducek on a rollout, we could consider resetting the policy status from the Policy Propagator or setting it to a new status indicating a rollout is in progress. The status sync could recognize this and reset the compliance status on the policy templates. There is already logic in the controllers to always send a compliance event when there is no status and it would cause configuration policies to be evaluated immediately regardless of evaluation interval. Then when the policies get a new status, we know the deployment has occurred.

You'd have to think about the specific workflow but I think that may be a good option to communicate between the hub and managed clusters during a rollout.

Author: Moving this thread to https://github.com/open-cluster-management-io/enhancements/pull/99#discussion_r1294997249 to prevent duplication.

>   - Update the `governance-policy-propagator` controller to add a `open-cluster-management.io/root-policy-generation`

Reviewer 4: Just thinking out loud, but I don't know if the generation number is enough here. Maybe I'm being paranoid, but if I delete the root policy and recreate it quickly enough that the propagator hasn't removed the old ones yet, I may have changed the policy while keeping the generation the same (or maybe the generation here is actually a larger than the one on the root policy). With the long-term compliance history stuff, we're already considering making a hash of the policy to identify it, I think? Maybe that could also be used here...

Author: I thought about the hash and especially if the compliance history would make use of it, that'd definitely be the way to go. It becomes complex, however, if hub templates are involved since the root policy and replicated policy would look different, which is why I went with the generation. If the root policy generation is less than the annotation on the replicated one, that could also be a signal to delete the replicated policy and recreate it.

    annotation to the replicated policy for each cluster.

>   - Update the annotation with the policy's generation at the time of the update when a policy has been replicated.

Reviewer 3: I am confused here. So this annotation is updated when a root policy has been replicated when the root policy changed e.g. pod limit change, Is this annotation changed?

Author: Yes, when the policy is replicated to the managed cluster namespace on the hub after the root policy is updated, the generation annotation will also be updated.

Reviewer 3: You mean when the root policy spec updated, policy is replicated, not update?

Author: Correct--there might be too many words in there--I'll update it.

  - (Optional--requirement for phase 3) Update the `governance-policy-propagator` controller to add a

>     `open-cluster-management.io/root-policy-timestamp` annotation to the replicated policy for each cluster indicating

Author: Correct--and that might be a better name for it 🙂

>     `open-cluster-management.io/root-policy-timestamp` annotation to the replicated policy for each cluster indicating

Reviewer 4: +1 on something like `open-cluster-management.io/replication-timestamp`

>     `open-cluster-management.io/root-policy-timestamp` annotation to the replicated policy for each cluster indicating

Author: How about `last-replicated`?

>     when the replicated policy was last applied.

Reviewer 2: As in the last time the managed cluster synced the propagated policy?

Author: The time that the governance-policy-propagator create/updated the replicated policy. That'd be the baseline for the timeout.

#### Story 2 / Phase 2

As a system administrator, I want to use a placement `DecisionStrategy` with policies.

- **Summary**

  - Use the `Placement` library to gather the list of clusters and iterate over them as it is today.

- **Details**

  - Make `Placement` the primary placement resource and parse `PlacementRule` decisions into the `Placement` struct
    instead of the `PlacementRule` struct.
  - Using the `Placement` library and implementing a basic `ClusterRolloutStatusFunc`, iterate over all clusters as
    usual using the `All` rollout strategy. The `RolloutStatus` will initially be either `ToApply` or `Succeed`.

>   - (Optional cleanup) Remove the current `Placement.Decisions` from the `status` of the `Policy` CRD. It's uncertain

Author: Right--I might make a separate issue to look at the CRD cleanup outside of this enhancement. But I'm also not sure it's used anywhere...

>   - (Optional cleanup) Remove the current `Placement.Decisions` from the `status` of the `Policy` CRD. It's uncertain

Reviewer 4: Re: API updates, I think we could stop populating the field, without removing it from the CRD (and therefore without requiring a version bump). Especially since it's just in the status, it doesn't feel like it's as much of a breaking change.

>   - (Optional cleanup) Remove the current `Placement.Decisions` from the `status` of the `Policy` CRD. It's uncertain

Author: @JustinKuli We'll have to investigate it (and I'll probably create a new issue to not muddle this enhancement), but I don't think we populate it currently 😅

    return clusterv1alpha1.NewRolloutHandler(pdTracker)
  }

  var GetClusterRolloutStatus clusterv1alpha1.ClusterRolloutStatusFunc = func(clusterName string) clusterv1alpha1.ClusterRolloutStatus {
    // (Phase 3 rollout logic goes here)

    return clusterv1alpha1.ClusterRolloutStatus{
      Status: "",             // initially determined by comparing the generation of the policy with the particular
                              //   cluster's generation, resolving to either "ToApply" or "Succeed"
      LastTransitionTime: "", // retrieved from the last transition time in the cluster's status
    }
  }
  ```

  ```golang
  rolloutHandler := common.GetRolloutHandler(client, placement)
  strategy, rolloutResult, err := rolloutHandler.GetRolloutCluster(clusterv1alpha1.All, common.GetClusterRolloutStatus)

  // ... Use rolloutResults for propagation
  ```

#### Story 3 / Phase 3 (Stretch goal)

As a system administrator, I want policies to be rolled out only when the policy on previous clusters show as
"Compliant" and control their concurrency and priority.

- **Summary**

>   - Only continue policy deployment once the previously deployed clusters show as "Compliant" after a configurable

Author: In the decision strategy, there's a timeout to wait for "Success" (which in the case of Policy is "Compliant"). The rollout will continue when clusters in the current rollout have succeeded OR have timed out.

    timeout.

- **Details**

  - Update the `PlacementBinding` CRD to contain the `RolloutStrategy` struct. (See the
    [`v1alpha1/RolloutStrategy`](https://github.com/open-cluster-management-io/api/blob/main/cluster/v1alpha1/types_rolloutstrategy.go))
    Default to `All` if a strategy is not provided.
  - Update the `governance-policy-propagator` controller to add a `open-cluster-management.io/root-policy-timestamp`
    annotation to the replicated policy for each cluster indicating when the replicated policy was last applied.

>   - Store a `lastEvaluated` in the status of the replicated policies so that the `RolloutStatus` can be determined

Reviewer 2: Could you please elaborate on this more?

I'm thinking about the case where the root policy is updated and the ConfigurationPolicy gets evaluated but the status doesn't change, how would this work? Also, if the root policy is updated but the generation of the ConfigurationPolicy doesn't change and its evaluation interval is like `2h` or `never` when compliant, this would not trigger an evaluation in time.

Other controllers also don't post when they last evaluated.

Author: We do provide the `LastEvaluated` in the ConfigurationPolicy, so I was thinking we can add logic to have the status-sync controller update the evaluation time when the `root-policy-timestamp` is greater than the last evaluated time, but I'd have to dig into when the `LastEvaluated` gets sent if it's not on every reconcile...
```golang
        // An ISO-8601 timestamp of the last time the policy was evaluated
        LastEvaluated string `json:"lastEvaluated,omitempty"`
```
https://github.com/open-cluster-management-io/config-policy-controller/blob/7e648171a8bd1ac051fe130f5539fbc36347907b/api/v1/configurationpolicy_types.go#L197 

Good point about other controllers, though. I'm not sure--I'll have to consider it more and look at our code. Maybe wipe the status on an update to let it repopulate or something?

Reviewer 2: @dhaiducek after brainstorming with @gparvin since he has a good sense of the requirements from a core user, I'd like to propose the following:

* New policies are progressively propagated to clusters regardless of whether they are inform or enforce. The status of the root policy should indicate that cluster in later groups are pending rollout or some kind of other state.
* Existing policies that are updated should only have their `remediationAction: enforce` progressively applied. Take for instance a scenario where we have clusters A, B, and C and A is a canary cluster. An update to my policy that sets `remediationAction: enforce` would at first have `remedation: enforce` on cluster A and `remediationAction: inform` on clusters B and C. The root policy status needs to indicate that clusters B and C are pending enforcement. For example, their compliance state could be `NonCompliant - Pending Rollout`.
* PolicyAutomation (separate story) would have a new mode of `everyRollout` so can be triggered when a rollout fails as opposed to each cluster becoming noncompliant.

@dhaiducek and @JustinKuli , what do you think?

Author: I like the idea to switch to inform to prevent unintended updates. We could add the `RolloutStatus` to each cluster's status, but I'm not sure we should change the compliance state to NonCompliant during a rollout, especially if users have alerts set up or are active users of the dashboards--that could get unintentionally noisy? Maybe we just use the `Pending` state? But overall I think this is the right direction to having a controlled rollout.

Reviewer 2: It's a fair point but at the same time, as a user, I would like to see the state of the policy on the cluster if there are inform policies applied even if they are pending rollout.

We could use the `Pending` state but include the noncompliant message so at least the information is there.

What do you think @dhaiducek and @gparvin?

Reviewer: I prefer the compliance being NonCompliant when that reflects reality and "maybe" have a separate Remediation Pending status (new field).  Though there may be too many things we could do in this extra remediation status field:
- pending rollout
- remediation applied by policy controller
- automated remediation applied
- remediated by automation or user intervention
- no remediation needed

Author: @gparvin I'm unclear on when you say "when that reflects reality". Are you saying policies should continue to report based on the content it has at the time, or are you saying that if the policy is out-of-date, it should be considered NonCompliant? For the latter, I suppose I'm thinking about a case where it might have been a minor update to the policy where the ultimate compliance wouldn't have changed due to the update? But maybe there's a "Policy" best practice I'm missing there if that's the meaning.
And, if we set it to "NonCompliant" because a rollout is happening, how would setting it to "NonCompliant" affect Ansible automation? As @gparvin points out, we might need some way to distinguish "NonCompliant--this is out of date" vs "NonCompliant--this is broken"? In that case, should Ansible automation only act when the `RolloutStatus` is `Succeed`? (I'm crossing my fingers that gets updated to `Succeeded` 😆)
Interestingly, to the "too many things we could do" point, having an `enum` for status is apparently against the Kubernetes status conventions (though it seems `RolloutStatus` is already set as such), where "Status" should simply be "True", "False", or "Unknown": https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties

Reviewer 2: @dhaiducek, I think the point is that if we follow the idea that a progressive rollout to a policy update makes the enforce part progressive and nothing else, then the pending rollout clusters should still accurately report back their inform status.

Having an additional field in the root policy status, propagated policy status, or both indicating that a rollout is pending would be sufficient in my opinion.

For the PolicyAutomation front, I don't want us to get caught up on this since it's a very underutilized feature. For now, we could have an additional optional spec field on the PolicyAutomation such as `ignorePendingRollout`.

Author: Okay, that all sounds good to me, too. I've updated the enhancement with a point about setting the `remediationAction` to `inform`. I still need to check whether removing the status might work as a better alternative to the `lastEvaluated` status.

    compliant (or after a configurable timeout):

    | Status        | Description                                                                                                                                          |
    | ------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------- |
    | `ToApply`     | No generation set                                                                                                                                    |
    | `Progressing` | Current generation and `lastEvaluated` has not been updated                                                                                          |
    | `Succeed`     | Compliant and current generation                                                                                                                     |
    | `Failed`      | NonCompliant and current generation                                                                                                                  |

>     | `TimeOut`     | Current generation, `lastEvaluated` is less than `root-policy-timestamp`, </br>and time has passed beyond timeout specified in the `RolloutStrategy` |

Author: The `lastEvaluated` might be reworked to use the transition time instead (See https://github.com/open-cluster-management-io/enhancements/pull/99#discussion_r1294678804). The root-policy-timestamp would be when the replicated policy spec was created/updated with the current generation, so any time before that would have been a previous version of the policy.

    | `Skip`        | (unused)                                                                                                                                             |

  - (Stretch) Update the `Policy` CRD to contain a `RolloutStatus`. This will be used to provide an aggregated rollout
    status for the root policy of `Progressing` (this would include `ToApply`), `Succeed`, or `Failed` (this would
    include `TimeOut`).

- **Snippet**

  ```golang
  type PlacementBinding struct {
    metav1.TypeMeta   `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`
    // +kubebuilder:validation:Optional
    BindingOverrides BindingOverrides `json:"bindingOverrides,omitempty"`
    // +kubebuilder:validation:Optional

>     RolloutStrategy clusterv1alpha1.RolloutStrategy `json:"rolloutStrategy,omitempty"` // <-- New field

Author: Good questions. I need to figure out how selective policy enforcement works, but I think they could work together, with the propagator filtering out clusters from the clusters returned from the Placement library.

The `additionalPrinterColumns` would be a good addition to the CRD!

>     RolloutStrategy clusterv1alpha1.RolloutStrategy `json:"rolloutStrategy,omitempty"` // <-- New field

Reviewer 2: @dhaiducek I worry about the rollout strategy being defined on the PlacementBinding instead of the Policy because you can have multiple PlacementBindings per Policy. If multiple set rolloutStrategy, which one wins? Could a subset of clusters rollout `Progressive` and others rollout `ProgressivePerGroup`?

Also, the PlacementBinding can be used for multiple policies and it'd be annoying to have to use a separate PlacementBinding just because your policies have different rollout strategies.

Lastly, I think the UX around PlacementBinding is challenging. I find it much more intuitive to configure how my policy is deployed by setting it on the policy.

>     RolloutStrategy clusterv1alpha1.RolloutStrategy `json:"rolloutStrategy,omitempty"` // <-- New field

Reviewer 2: In regards to selective policy enforcement, I personally think it should be ignored or an error returned when a rollout strategy is also defined.

>     RolloutStrategy clusterv1alpha1.RolloutStrategy `json:"rolloutStrategy,omitempty"` // <-- New field

Reviewer 5: The multiple placement bindings question also seems to apply to the order/timing of when the enforce is done. If two bindings for the same policy have overlapping sets of clusters included both will (as I understand it) calculate decisiongroups independently. Any given cluster may be in group 1 of binding A and group 100 of binding B. Will it get enforced twice? If the max concurrency is 10, do you end up with 11 clusters being enforced due to the different groupings? etc.

    // This field provides the ability to select a subset of bound clusters
    // +kubebuilder:validation:Optional
    // +kubebuilder:validation:Enum=restricted
    SubFilter SubFilter `json:"subFilter,omitempty"`
    // +kubebuilder:validation:Required
    PlacementRef PlacementSubject `json:"placementRef"`
    // +kubebuilder:validation:Required
    // +kubebuilder:validation:MinItems=1
    Subjects []Subject              `json:"subjects"`
    Status   PlacementBindingStatus `json:"status,omitempty"`
  }
  ```

  ```golang
  // ComplianceHistory defines compliance details history
  type ComplianceHistory struct {

>     LastEvaluated metav1.Time `json:"lastEvaluated,omitempty" protobuf:"bytes,7,opt,name=lastEvaluated"`  // <-- New field

Reviewer 2: Similar to my last comment, how is this field set?

Author: Moving thread to https://github.com/open-cluster-management-io/enhancements/pull/99#discussion_r1294997249

    LastTimestamp metav1.Time `json:"lastTimestamp,omitempty" protobuf:"bytes,7,opt,name=lastTimestamp"`
    Message       string      `json:"message,omitempty" protobuf:"bytes,4,opt,name=message"`
    EventName     string      `json:"eventName,omitempty"`
  }

  // PolicyStatus defines the observed state of Policy
  type PolicyStatus struct {
    Placement []*Placement                  `json:"placement,omitempty"`
    Status    []*CompliancePerClusterStatus `json:"status,omitempty"`

    // +kubebuilder:validation:Enum=Compliant;Pending;NonCompliant
    ComplianceState ComplianceState               `json:"compliant,omitempty"`
    RolloutStatus   clusterv1alpha1.RolloutStatus `json:"rolloutStatus,omitempty"` // <-- New field
    Details         []*DetailsPerTemplate         `json:"details,omitempty"`
  }
  ```

### Implementation Details/Notes/Constraints

For the `Placement` library, this requires importing at least this package version (the `Placement` library is in the
`v1alpha1` version):

```
  open-cluster-management.io/api v0.11.1-0.20230809113658-2b2399b5f6e8
```

> For testing, the `governance-policy-propagator` doesn't currently account for multiple managed clusters. As part of this

Reviewer 4: I think it does currently consider multiple clusters, just in a fake way (which is probably what we want). I'm not saying that what it does is necessarily great, it probably needs to be looked at in detail for testing this feature, but it is already there to an extent: https://github.com/stolostron/governance-policy-propagator/blob/main/test/e2e/case1_propagation_test.go#L78

  for implementation. The phased approach intends to address this to make partial implementation feasible.
- Having a `LastEvaluated` field in the status is going to increase load on the hub cluster Kubernetes API server that
  efforts originally sought to reduce by not sending updates when the status hadn't changed. This could be remedied by
  only sending an updated timestamp if the generation was updated and/or the `LastEvaluated` timestamp is less than the
  root policy timstamp or (the current behavior) if the status indeed changed.

## Design Details

### Open Questions

1. Should the rollout handler be instantiated with the dynamic watcher instead of the controller client?
2. Should `RolloutStatus` be reflected per-cluster in the root policy? (Removing `clusterNamespace` could essentially
   make the CRD the same size as it currently is, remediating concerns around bloat, but the rollout status is also
   reflected in each replicated policy.)
3. Is there opposition to removing `Placement.Decisions` from the `status`? (This could alternatively be handled outside
   of this enhancement.)

> 4. Should the `RolloutStrategy` be implemented in the `Policy` CRD instead of `PlacementBinding`?

Reviewer 2: I prefer this level.

Reviewer: This means it needs to be in PolicySet as well.

Author: @mprahl I just realized that "this" was unclear. Are you saying you prefer it being in the `Policy`?

5. How should a policy handle when placement decision groups shift? (This implementation leans on the `Placement`
   library to handle such shifts, only handling the clusters that it returns.)
6. (An aside) What's with the pointer in `[]*clusterv1beta1.PlacementDecision`?

> 

Author: Right, I'm not sure placement impacts PolicyAutomation, but definitely something to verify.

Odd indeed, but the old templates ought to be reflected in the replicated policy, and it'd hopefully be relatively straightforward to find the rollout status in the root policy to figure out that it hadn't been updated yet. That said, maybe storing a per-cluster rollout status in the root policy would be good for visibility.

been bound. The new `DecisionStrategy` field in the `Placement` API creates cluster groupings that controllers can
leverage to have segmented, configurable rollouts of resources. This will aid use cases where high availability may be a
priority or a set of test clusters should receive and verify the updates before the remaining clusters get updated.

### Goals

- Make `Placement` the primary API for placement (currently the governance propagator is somewhat
  `PlacementRule`-centric)
- Leverage the `Placement` helper library to retrieve placement decisions
- Reflect rollout status per cluster in the root policy for discoverability (whether a cluster is up-to-date or not)
- (stretch) Implement the `RolloutStrategy` struct for policies, including:
  - `RolloutStrategy`
    - "All": all clusters at once
    - "Progressive": one cluster at a time
    - "ProgressivePerGroup": one group of clusters at a time
  - `Timeout`: Maximum amount of time to wait for success before continuing the rollout
  - `MandatoryDecisionGroups`: groups that should be handled before other groups
  - `MaxConcurrency`: Concurrency during a progressive rollout

> - (stretch) Add an aggregated rollout status for the root policy status.

Reviewer 2: ```suggestion
- Add an aggregated rollout status for the root policy status.
```

clusters.

## Motivation

Currently policies and subsequent updates are pushed out to clusters en masse based on the placement to which it has
been bound. The new `DecisionStrategy` field in the `Placement` API creates cluster groupings that controllers can
leverage to have segmented, configurable rollouts of resources. This will aid use cases where high availability may be a
priority or a set of test clusters should receive and verify the updates before the remaining clusters get updated.

### Goals

- Make `Placement` the primary API for placement (currently the governance propagator is somewhat
  `PlacementRule`-centric)
- Leverage the `Placement` helper library to retrieve placement decisions
- Reflect rollout status per cluster in the root policy for discoverability (whether a cluster is up-to-date or not)
- Implement the `RolloutStrategy` struct for policies, including:
  - `RolloutStrategy`
    - "All": all clusters at once
    - "Progressive": one cluster at a time
    - "ProgressivePerGroup": one group of clusters at a time
  - `Timeout`: Maximum amount of time to wait for success before continuing the rollout

>   - `MandatoryDecisionGroups`: groups that should be handled before other groups

Author: This would solely be priority ordering, but that's an interesting point to have a different timeout setting for the mandatory groups and the subsequent groups--that could be an update for the Placement library in the future.

`MandatoryDecisionGroups` is described in the `DecisionStrategy` enhancement here: https://github.com/open-cluster-management-io/enhancements/tree/main/enhancements/sig-architecture/64-placementStrategy#how-workload-applier-apis-will-benefits-from-using-placementstrategy

  - `MaxConcurrency`: Concurrency during a progressive rollout
- Add an aggregated rollout status for the root policy status.

### Non-Goals

Any specialized behaviors outside of those provided by the `Placement` library and, by extension, the `DecisionStrategy`
enhancement, that might require additional code other than that already provided/required by the `Placement` library.
This includes the ability to roll back a rollout. A rollback should be implemented by the user by applying the previous

> version of the policy, and GitOps is the recommended method for this.

Author: @imiller0 That's true for objects on the managed cluster--this is referring specifically to the previous version of the policy itself on the hub rather than rolling back changes on the managed cluster.

## Proposal

### Background

Users create policies on the hub cluster (which are referred to as the root policy). These policies are replicated to
managed cluster namespaces based on the placement to which they have been bound by the associated `PlacementBinding`
object. The managed clusters monitor the associated managed cluster namespace on the hub for policy updates.

Additionally, policies have a `remediationAction` specified, which can either be `inform`, meaning the policy controller
monitors objects on the managed cluster reporting status without making changes, or `enforce`, meaning the policy
controller attempts to make changes on the managed cluster based on the policy definition and reports status accordingly
based on the success or failure of the update.

### Design

For this rollout enhancement, policy propagation by the `governance-policy-propagator` controller on the hub would take
place in a multiple-pass approach:

1. On the first pass following a root policy creation or update, the `governance-policy-propagator` controller on the
   hub cluster will replicate all policies to managed cluster namespaces and are created regardless of the rollout

>    strategy specified, with the remediation action set to `inform`. This way, all policies are up-to-date and reporting

Reviewer 5: If a policy is modified/updated by the user while a rollout is ongoing, does the process start over with all policies being set back to inform and rollout beginning with group 1 again?

Author: That's correct

   a compliance status based on the current version of the root policy. The aggregated rollout status on the root policy
   is set to `Progressing`. If the remediation action on the root policy is already `inform`, the rollout status on each

>    cluster is set to `Progressing` and no second pass occurs. If the remediation action is `enforce`, the rollout status

Reviewer: Does an inform policy that rolls out with some clusters being NonCompliant ever transition out of Progressing to Succeeded?  I think a clarification is needed on how an inform policy rollout can be successful and still identify NonCompliant clusters.

Author: It's an interesting question. I was thinking `inform` would follow the `enforce` behaviors, transitioning to `Succeeded` or `Failed` based on the compliance.

But for `inform`, I suppose we could transition the rollout to `Succeeded` once a status is returned regardless of compliance? It'd differ in behavior from `enforce`, but maybe that's okay?

Reviewer: Yes, for inform I think as customers learn where problems "may" be they will improve inform policies to detect those issues.  Then the rollout will help identify those problem clusters and I don't think the rollout should fail or be limited because inform policies are working as expected.  Example: Initially we want our addon to just be available.  In v2 we learn we want to look out for terminating config policies....

Reviewer 2: A user shouldn't define a progressive rollout for an inform policy since it will cause the policy to be enforced unless the ConfigurationPolicy has `remediationAction: informOnly` which in that event, the rollout would have no effect.

I think the rollout status should be set to `Succeeded` if any status is provided as Dale mentioned. Syntax errors in the ConfigurationPolicy would lead to a failed rollout status.

Reviewer 2: The nice part of this is that even users that don't use progressive policy rollouts will get a single field to know if all policy templates have reported an up to date status for the cluster.

Reviewer: Can I not roll out a policyset that contains inform and enforce policies and the inform policies would not be enforced? I should be able to put any random policies in a policyset with some enforce and some inform and it should be honored appropriately during rollout.  Enforcing some policies could have unintended consequences -- rebooting nodes, upgrading, etc

Reviewer 6: I think it needs more clarification on how the RolloutStatus will move from ToApply->progressing->Succeeded/TimeOut. 
Here is some thoughts;
- **ToApply;** 
     1- when the child policy does not get created yet for the managedCluster
     2- when the child policy is created for the managedCluster **AND** the policy remediation is Inform and the policy in NonCompliant state.
      
- **Progressing;** 
    1- When the child policy is created for the managedCluster **AND** the policy remediation is Enforce and the policy in NonCompliant state.
      
- **Succeeded;** 
   1- When the child policy is created for the managedCluster and it is in Compliant state for either enforce/inform remediation

- **TimeOut;** 
   1- When the child policy is created for the managedCluster  **AND** the policy remediation is Enforce and the policy in NonCompliant state for longer time than it is defined in the rolloutStrategy->TimeOut

   on each cluster is set to `ToApply`.
2. On the subsequent passes, if the `remediationAction` is `enforce`, the `governance-policy-propagator` fetches the
   managed clusters returned from the `Placement` library (based on the user-configured rollout strategy) and sets the
   rollout status of each returned cluster to `Progressing`, updating any `remediationAction` to `enforce` as defined in
   the root policy.

From here, work is picked up by the `governance-policy-framework` controllers on the managed clusters:

1. The template sync controller of the framework updates the applicable `ConfigurationPolicy` objects like usual, which
   triggers reevaluation from the `config-policy-controller`.
2. Once the `ConfigurationPolicies` have a generation that matches the `lastEvaluatedGeneration` set on the policy

>    status, the status sync controller of the framework will set the rollout status to to `Succeeded` after all of the

Reviewer 2: You should mention that this is set on the replicated policy on the managed cluster and hub.

>    `ConfigurationPolicies` have a compliant status or `Failed` if there is any non-compliant status.

Reviewer 2: You should mention that the propagator will add the rollout status on the root policy after the propagated policy rollout status is updated.

### User Stories

#### Story 1

As a system administrator, I want to know the status of a rollout for a policy and what clusters have been updated.

- **Summary**

  - Add a `RolloutStatus` to the `Policy` status in the CRD.

- **Details**

  - The `RolloutStatus` would be added to reflect: rollout status on the replicated policy, per-cluster on the root
    policy, and an aggregated status on the root policy. The aggregated status would specify: `Progressing` (this would

>     include `ToApply`), `Succeeded`, `Failed`, or (this would include `TimeOut`).

Reviewer 2: I think `ToApply` means the the policy hasn't been enforced yet on the cluster.

`Progressing` would mean the policy has been enforced on the cluster but the rollout hasn't completed yet.

Reviewer 5: Is the set of "valid" status fields different on the root policy vs the replicated policies? It sounds like ToApply would be valid on a replicated policy but not the root. Similarly it sounds like progressing would be valid on the root but not the replicated set. If so it might help to define the set of valid values for root vs replicated.

Author: Good point, @mprahl--I wrapped `ToApply` in for simplicity but it would be clearer to use `ToApply`.

@imiller0 The `RolloutStatus` is an iota we're taking from the Placement library: https://github.com/open-cluster-management-io/api/blob/9675ab737d2086dd7094fc962c9617cc71ec7484/cluster/v1alpha1/helpers.go#L21-L38 but having these well defined for each the root and replicated policies would be good.

>     include `ToApply`), `Succeeded`, `Failed`, or (this would include `TimeOut`).

Reviewer 6: I don't think there is a Failed status cause the policy-controller will continue to enforce the policy to the managedCluster as long as it is in NonCompliant state. So status could be  **ToApply, Progressing, Succeeded, TimeOut.**

Author: @serngawy True, but if the enforcement is blocked for some reason (like a missing API on the managed cluster or an invalid manifest), it may be good to reflect that in the root policy since there's no way for the rollout to progress?

    _, _, err := pdTracker.Get()
    if err != nil {
      log.Error(err, "Error retrieving PlacementDecisions from tracker")
    }

    return clusterv1alpha1.NewRolloutHandler(pdTracker)
  }

  var GetClusterRolloutStatus clusterv1alpha1.ClusterRolloutStatusFunc = func(clusterName string) clusterv1alpha1.ClusterRolloutStatus {
    // Fetch and return the rollout status and the last transition time from the status in the replicated policy

    return clusterv1alpha1.ClusterRolloutStatus{
      Status: "<rollout-status>",
      LastTransitionTime: "<transition-time>",
    }
  }
  ```

  ```golang
  rolloutHandler := common.GetRolloutHandler(client, placement)
  strategy, rolloutResult, err := rolloutHandler.GetRolloutCluster(clusterv1alpha1.All, common.GetClusterRolloutStatus)

  // ... Use rolloutResults for propagation
  ```

#### Story 3

As a system administrator, I want policies to be rolled out only when the policy on previous clusters show as
"Compliant" and control their concurrency and priority.

- **Summary**

  - Only continue policy deployment once the previously deployed clusters show as "Compliant".

- **Details**

>   - Update the `PlacementBinding` CRD to contain the `RolloutStrategy` struct. (See the

Reviewer 6: I think its better to define the RolloutStrategy under the Policy->Spec not under the PlacementBinding cause a policy can have multiple placementBinding so as a user should I repeated RolloutStrategy across all placementBindings? 
I guess defining the RolloutStrategy under PlacementBinding  will be challenging to handle.

Author: I was working on updating this as you were reviewing 🙂 We came to the same conclusion and have updated the enhancement to specify the RolloutStrategy on the Policy rather than the PlacementBinding.

    [`v1alpha1/RolloutStrategy`](https://github.com/open-cluster-management-io/api/blob/main/cluster/v1alpha1/types_rolloutstrategy.go))
    Defaults to `All` if a strategy is not provided or the remediation action is `inform`.

>   - When the `remediationAction` is set to `enforce`, policies not currently being rolled out will be set to `inform` to

Reviewer 5: Once set to enforce do they remain in enforce mode until a subsequent change to the root policy or do they revert to inform once compliant?

Author: They remain in `enforce` until the root policy is updated again.

`v1alpha1` version):

```
  open-cluster-management.io/api v0.11.1-0.20230828015110-b39eb9026c6e
```

For testing, the `governance-policy-propagator` doesn't currently account for multiple managed clusters. As part of this
enhancement, the test flows would need to be enhanced (and/or a separate workflow created) that deploys multiple managed
clusters.

The `PlacementDecisionGetter` returns an array of pointers (`[]*clusterv1beta1.PlacementDecision`) because it was
intended to retrieve from a cache, so the implementation could consider setting up a `PlacementDecision` cache instead
of using the Kubernetes client directly.

### Risks and Mitigation

- The `Placement` library is relatively new and untested outside of its repo and this implementation leans heavily on
  its logic. While it works in theory, there could be some tweaks/adjustments as development proceeds, lengthening time
  for implementation. The phased approach intends to address this to make partial implementation feasible.

## Design Details

### Open Questions

1. Should the rollout handler be instantiated with a dynamic watcher/cache instead of the controller client?

> 2. Should `RolloutStatus` be reflected per-cluster in the root policy? (The rollout status would also be reflected in

Reviewer 2: Yes to this.

   each replicated policy.)
3. Should the `RolloutStrategy` be implemented in the `Policy` CRD instead of `PlacementBinding`? (If that were the
   case, it would also need to be added to `PolicySet`.)
4. How should a policy handle when placement decision groups shift? (This implementation leans on the `Placement`

>    library to handle such shifts, only handling the clusters that it returns.)

Reviewer 5: Does this risk clusters being skipped? Or gaps in the decision groups if clusters are unbound/rebound?

> 5. How is "soak time" (the minimum time for a policy to reach a successful state) and timeout (the maximum amount of
>    time to wait for a successful state), as defined in the placement enhancement, handled for Policies?

Reviewer 2: ```suggestion
5. How is "soak time" (the minimum time for a policy to reach a successful state) and timeout (the maximum amount of
   time to wait for a state), as defined in the placement enhancement, handled for Policies?
```

> 

Reviewer 5: How does the rolloutStrategy work with policy dependency? Is it possible that if the two policies are bound using separate policybindings and the policies are modified simultaneously the group ordering could be such that rollout of one policy is entirely blocked until completion of all groups of the other policy?

### Design

Upon policy creation or update, all clusters receive the same version, preventing version skew among different clusters.
However, the `remediationAction` for enforced policies is initially set to `inform` and the rollout occurs as each
cluster in turn has its policy updated to `enforce`. For each rollout strategy, this looks like:

| Rollout Strategy      | Remediation Inform                      | Remediation Enforce                                                                                                                      |
| --------------------- | --------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------- |
| `All`                 | Policy rolled out to all clusters as-is | Policy rolled out to all clusters as-is                                                                                                  |
| `Progressive`         | (Same as `All`)                         | Policy applied initially to all clusters as `inform` and <br/>rollout enforces each cluster's policy once compliant (or timed out)       |

> | `ProgressivePerGroup` | (Same as `All`)                         | Policy applied initially to all clusters as `inform` and <br/>rollout enforces each cluster group's policy once compliant (or timed out) |

Reviewer 2: ```suggestion
| `ProgressivePerGroup` | (Same as `All`)                         | Policy applied initially to all clusters as `inform` and <br/>rollout enforces a cluster group's policy and waits for compliant (or timed out) before moving to the next group |
```

  ```yaml
  apiVersion: policy.open-cluster-management.io/v1
  kind: Policy
  metadata:
    name: sample-policy
    namespace: policy-namespace
  spec:
    disabled: false
    remediationAction: enforce
    rolloutStrategy:               # <-- New field
      type: ProgressivePerGroup
    policy-templates:
      - objectDefinition:
          apiVersion: policy.open-cluster-management.io/v1
          kind: ConfigurationPolicy
          ...
  ---
  apiVersion: policy.open-cluster-management.io/v1
  kind: PlacementBinding
  metadata:
    name: sample-placementbinding
    namespace: policy-namespace
  placementRef:
    name: sample-placement
    apiGroup: cluster.open-cluster-management.io
    kind: Placement
  subjects:
    - name: sample-policy
      apiGroup: policy.open-cluster-management.io
      kind: Policy
  ```

  **Replicated policy in managed cluster namespace**

  ```yaml
  apiVersion: policy.open-cluster-management.io/v1
  kind: Policy
  metadata:
    name: policy-namespace.sample-policy
    namespace: cluster1
  spec:
    disabled: false
    remediationAction: inform # <-- Remediation action updated to prepare for rollout
    rolloutStrategy:

>       type: ProgressivePerGroup

Reviewer 6: I guess better to include the progressivePerGroup fields as well.
```disabled: false
remediationAction: inform # <-- Remediation action updated to prepare for rollout
rolloutStrategy:
  type: ProgressivePerGroup
  progressivePerGroup:
    mandatoryDecisionGroups:
    - groupName: dev-emea   (Same group name as defined in the placement.)
    timeout: None```
